opentelemetry/proto/metrics/v1/metrics.proto

```diff
@@ -95,14 +95,103 @@ message InstrumentationLibraryMetrics {
+//
+// Data points with/ the 0 value for TimeUnixNano SHOULD be rejected
+// by consumers.
+//
+// # StartTimeUnixNano
+//
+// This field indicates to consumers the start time for points with
+// cumulative and delta AggregationTemporality, and it should be
+// included whenever possible to support correct rate calculation and
+// overlap detection.  Although it may be omitted when the start time
+// is truly unknown, setting StartTimeUnixNano is strongly encouraged.
+//
+// There are two reasons that streams with AggregationTemporality are
+// reset: (1) a process restarts, and (2) the SDK "forgets" the stream
+// and then re-initializes it.  StartTimeUnixNano is critical for
+// calculating correct rates for Monotonic Sum, Histogram, and Summary
+// data points in both of these cases
+//
+// StartTimeUnixNano may be used with Gauge instruments to indicate
+// when the measurement was captured, as opposed to the moment it was
+// observed.
+//
+// Another use for this information is to prevent overlap when
+// multiple writers accidentally write to a single stream--applies
+// to all data points.
+//
+// # Unknown StartTimeUnixNano
+//
+// This protocol adopts the Prometheus heuristic for reset detection,
+// and with it prescribes a solution for filling in StartTimeUnixNanos
+// with as much information as possible to assist consumers with
+// calculating rates and detecting overlap.
+//
+// The Prometheus heuristic indicates that a reset must have occured
+// in a monotonic series whenever the value descends from one value to
+// the next, in a timestamp-ordered sequence.
+//
+// To facilitate reset and overlap detection, when single-timestamped
+// sample values are translated into OTLP, such as from the Prometheus
+// Remote-Write protocol, use the following.
+//
+// ## Filling in StartTimeUnixNano
+//
+// To import a sequence of non-overlapping measurements into into OTLP
+// using a stateful processor:
+//
+// 1. Keep a map, using the metric name and attributes as the key, containing
+//    a struct of three parts:
+//    - resetTimestamp: the first observation or the reset timestamp
+//    - previousTimestamp: the timestamp of the most recent observation
+//    - previousValue: a monotonic value
+// 2. When a new point arrives and it is the first observation:
+//    - in the map, set resetTimestamp to the input timestamp
+//    - in the map, set previousValue to the input value
+//    - in the map, set previousTimestamp to resetTimestamp
+//    - in the output, set StartTimeUnixNano to resetTimestamp
+// 3. For each subsequent point that arrives in the sequence:
+//    - if the sample value is less than previousValue,
+//      then set resetTimestamp to previousTimestamp
+//    - set previousValue to the input value
+//    - set previousTimestamp to the input timestamp
+//    - in the output, set StartTimeUnixNano to resetTimestamp
+//
+// This technique ensures that all output points have a meaningful
```

Comment:
Just so that I understand: this assumes that the input points arrive in the order they are observed and no points are lost, right? If these conditions are violated we may have an incorrect output and that is an expected outcome, we don't try to solve this problem.

Reply:
Yes. There is a certain amount of ambiguity in this process, by nature. In the Prometheus setting, restarts are impossible to detect without this heuristic, and it is just a heuristic. 

The intention here is to say that when a new processor starts observing points, it must assume that earlier points (prior to the start of observation) are lost or previously recorded, whereas we are saying it is OK to presume that no resets actually take place between observations after the initial observation. 

It's OK to continue losing points in the input, but there will still be reset ambiguity.